File: frontend/src/types/students/student.ts
import type { ActiveFlag } from '@app-types/settings/session';

export interface StudentListItem {
  id: number;
  admission_no: string;
  roll_no: number | null;
  firstname: string;
  middlename: string | null;
  lastname: string | null;
  full_name: string;
  gender: string | null;
  mobileno: string | null;
  email: string | null;
  dob: string | null;
  is_active: ActiveFlag;
  class_id: number | null;
  section_id: number | null;
  class_name: string | null;
  section_name: string | null;
  admission_date: string | null;
  created_at: string;
  disable_reason_id?: number | null;
  disable_reason_name?: string | null;
  disable_note?: string | null;
  disabled_at?: string | null;
}

export interface StudentDetail extends StudentListItem {
  father_name: string | null;
  mother_name: string | null;
  guardian_name: string | null;
  guardian_phone: string | null;
  current_address: string | null;
  permanent_address: string | null;
  blood_group: string | null;
  religion: string | null;
  category_id: string | null;
  rte: string | null;
  updated_at: string | null;
}

export interface CreateStudentPayload {
  admission_no: string;
  admission_date: string;
  firstname: string;
  middlename: string | null;
  lastname: string;
  gender: string;
  dob: string;
  class_id: number;
  section_id: number;
  roll_no: number | null;
  mobileno: string | null;
  email: string | null;
  father_name: string | null;
  mother_name: string | null;
  guardian_name: string | null;
  guardian_phone: string | null;
  current_address: string | null;
  blood_group: string | null;
  religion: string | null;
  category_id: string | null;
  rte: string;
  is_active: ActiveFlag;
}

export type UpdateStudentPayload = CreateStudentPayload;

File: backend/apps/students/api/serializers/student.py
from rest_framework import serializers


class StudentCreateSerializer(serializers.Serializer):
    admission_no = serializers.CharField(max_length=100)
    admission_date = serializers.CharField(required=False, allow_blank=True)
    roll_no = serializers.IntegerField(required=False, allow_null=True)
    firstname = serializers.CharField(max_length=100)
    middlename = serializers.CharField(
        required=False, allow_blank=True, allow_null=True, max_length=255
    )
    lastname = serializers.CharField(max_length=100)
    gender = serializers.CharField(required=False, allow_blank=True, max_length=100)
    mobileno = serializers.CharField(required=False, allow_blank=True, allow_null=True)
    email = serializers.EmailField(required=False, allow_blank=True, allow_null=True)
    dob = serializers.CharField(required=False, allow_blank=True)
    class_id = serializers.IntegerField()
    section_id = serializers.IntegerField()
    father_name = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    mother_name = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    guardian_name = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    guardian_phone = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    current_address = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    blood_group = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    religion = serializers.CharField(required=False, allow_blank=True, allow_null=True)
    category_id = serializers.CharField(
        required=False, allow_blank=True, allow_null=True
    )
    rte = serializers.CharField(required=False, allow_blank=True)
    is_active = serializers.CharField(required=False, allow_blank=True)


class StudentUpdateSerializer(StudentCreateSerializer):
    admission_no = serializers.CharField(
        required=False, allow_blank=True, max_length=100
    )
    firstname = serializers.CharField(required=False, allow_blank=True, max_length=100)
    lastname = serializers.CharField(required=False, allow_blank=True, max_length=100)
    class_id = serializers.IntegerField(required=False)
    section_id = serializers.IntegerField(required=False)
